js/web/lib/wasm/jsep/webnn/buffer-manager.ts

```diff
@@ -0,0 +1,223 @@
+// WebNN API specification.
+// https://github.com/webmachinelearning/webnn/issues/677
+/// <reference path="webnn.d.ts" />
+
+export type BufferId = number;
+
+/**
+ * Manages BufferId to MLBuffer mapping.
+ */
+export interface BufferManager {
+  /**
+   * Reserve a new BufferId.
+   */
+  reserveBufferId(): BufferId;
+  /**
+   * Release a BufferId.
+   */
+  releaseBufferId(bufferId: BufferId): void;
+  /**
+   * Get MLBuffer by BufferId.
+   */
+  getBuffer(bufferId: BufferId): MLBuffer;
+  /**
+   * Ensure a MLBuffer is created for the BufferId.
+   */
+  ensureBuffer(bufferId: BufferId, dataType: MLOperandDataType, dimensions: number[]): MLBuffer;
+  /**
+   * Upload data to a MLBuffer.
+   */
+  upload(bufferId: BufferId, data: Uint8Array): void;
+  /**
+   * Download data from a MLBuffer.
+   */
+  download(bufferId: BufferId): Promise<ArrayBuffer>;
+  /**
+   * Release all buffers for a MLContext.
+   */
+  releaseBuffersForContext(mlContext: MLContext): void;
+  /**
+   * Register an externally created MLBuffer with a given MLContext and return a BufferId.
+   */
+  registerBuffer(mlContext: MLContext, mlBuffer: MLBuffer): BufferId;
+}
+
+let bufferGuid = 1;
+const createNewBufferId = (): BufferId => bufferGuid++;
+
+/**
+ * BufferTracker tracks the MLBuffer and pending upload data.
+ *
+ * We need to track the MLBuffer and pending upload data because we delay the creation of MLBuffer until
+ * we know the data type and dimensions. This is because future implementations of WebNN will only support creating
+ * MLBuffers with dataTypes and dimensions.
+ */
+class BufferTracker {
+  private mlBuffer?: MLBuffer;
+  private activeUpload?: Uint8Array;
+
+  constructor(private mlContext?: MLContext, buffer?: MLBuffer) {
+    this.mlBuffer = buffer;
+  }
+
+  public get buffer(): MLBuffer|undefined {
+    return this.mlBuffer;
+  }
+
+  public get context(): MLContext {
+    if (!this.mlContext) {
+      throw new Error('MLContext has not been set.');
+    }
+    return this.mlContext;
+  }
+
+  public set context(mlContext: MLContext) {
+    if (this.mlContext && this.mlContext !== mlContext) {
+      throw new Error('MLBuffer in use in a different MLContext.');
+    }
+    this.mlContext = mlContext;
+  }
+
+  public destroy(): void {
+    this.mlBuffer?.destroy();
+    this.mlBuffer = undefined;
+  }
+
+  public ensureBuffer(dataType: MLOperandDataType, dimensions: number[]): MLBuffer {
+    if (this.mlBuffer) {
+      return this.mlBuffer;
+    }
+
+    const buffer = this.context.createBuffer({dataType, dimensions});
+    this.mlBuffer = buffer;
+
+    if (this.activeUpload) {
+      this.mlContext?.writeBuffer(buffer, this.activeUpload);
+      this.activeUpload = undefined;
+    }
+
+    return buffer;
+  }
+
+  public upload(data: Uint8Array): void {
+    if (!this.mlBuffer) {
+      this.activeUpload = new Uint8Array(data);
```

Comment:
Could we just refer to the `data` instead of making a new copy? So we can avoid the extra copy of initializers on CPU.
```suggestion
      this.activeUpload = data;
```

Reply:
As the offset into the wasm heap could have changed or freed between `uploadBuffer` and `ensureBuffer`/buffer creation, we have to create a copy. The change above should have fixed this issue.